src/plugins/expressions/common/ast/ast_builder.test.ts

```diff
@@ -0,0 +1,136 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { astBuilder } from './ast_builder';
+
+describe('ast builder', () => {
+  test('correctly builds the AST', () => {
+    const exprAST = astBuilder.createExpression([
+      astBuilder.createFunction('test1', {
+        arg1: true,
+        arg2: 'test',
+        arg3: 4,
+        arg4: astBuilder.createExpression([
+          astBuilder.createFunction('subexpr_funct', {
+            arg1: true,
+          }),
+        ]),
+      }),
+      astBuilder.createFunction('test2', {
+        arg1: true,
+        arg2: 'test',
+      }),
+      astBuilder.createFunction('test3', {
+        arg1: true,
+      }),
+    ]);
+
+    expect(exprAST).toMatchSnapshot();
+  });
+
+  test('correctly builds the AST using builder functions', () => {
+    const ast = astBuilder
+      .createExpressionBuilder([
+        astBuilder
+          .createFunctionBuilder('test1', {
+            arg1: [true, false],
+            arg2: astBuilder
+              .createExpressionBuilder()
+              .addFunction(
+                astBuilder
+                  .createFunctionBuilder('test2', {
+                    arg1: true,
+                  })
+                  .toAst()
+              )
+              .toAst(),
+          })
+          .toAst(),
+      ])
+      .addFunction(astBuilder.createFunctionBuilder('test3').toAst())
+      .toAst();
+
+    expect(ast).toMatchSnapshot();
+  });
+
+  test('correctly builds the AST using combination of builders', () => {
+    const ast = astBuilder.createExpressionBuilder();
+    ast.addFunction(
+      astBuilder.createFunction('test1', {
+        arg1: [true, false],
+        arg2: astBuilder.createExpression([astBuilder.createFunction('test2')]),
+      })
+    );
+    const fn = astBuilder.createFunctionBuilder('test3');
+    fn.addArgument('arg1', true);
+    fn.addArgument(
+      'arg2',
+      astBuilder.createExpression([
+        astBuilder
+          .createFunctionBuilder('test4')
+          .addArgument('arg1', true)
+          .toAst(),
+      ])
+    );
+    ast.addFunction(fn.toAst());
+
+    expect(ast.toAst()).toMatchSnapshot();
+  });
+
+  test('best way to use it', () => {
+    const {
+      createExpressionBuilder,
+      createFunctionBuilder,
+      createFunction,
+      createExpression,
+    } = astBuilder;
+
+    // if you know everything about expression upfront use createExpression
+    const ast = createExpression([createFunction('test', { arg1: true })]);
```

Comment:
I wonder if having separate `createX` and `createXBuilder` functions makes this more confusing.

Thinking out loud -- What if we only provided the builders, and at the end when you call `ExpressionBuilder.toAst()` it goes through all of the functions & calls `.toAst()` on each of them?

That way you could hang on to references to functions even after they've been added to the expression, and make updates.

For example (I'm using a class-based approach for the builders here, but you get the idea):

```js
import { expressionAstBuilder } from '../expressions/public';
const { Expression, ExpressionFunction } = expressionAstBuilder;

// create instance of expression builder with one function as the "initial state"
const myExpression = new Expression([
  new ExpressionFunction('esaggs', { index: 'abc123' }
]);

// add another function
const fn = new ExpressionFunction('foo', { bar: true });
myExpression.addFunction(fn);

// update the function you've just added
const subExpression = new Expression([...]);
// works cause you saved a reference to it
fn.addArgument('bar', subExpression);

// toAst will go through and call toAst on each of the provided functions
const ast = myExpression.toAst();

// I suppose you could also offer a way to get the string as a convenience.
// It would call toAst() and then convert it to the expression string.
const string = myExpression.toString();
```